### PR TITLE
Create model

### PR DESCRIPTION
**Pull request by Redevic**

### CODE REVIEW

File: src/model/models.py [hunk 1]
```diff
@@ -0,0 +1,235 @@
...
+            count_windows (tp.Optional[int], optional): Count windows for attention. Defaults to 4.
+        """
+        super().__init__()
+
+        efficientnet = torchvision.models.efficientnet_b0(pretrained=True)  # pretrained efficientnet_b0
+        
+
+        # Remove linear and pool layers (since we're not doing classification)
+        modules = list(efficientnet.children())[:-2]
+        self.cvnet = nn.Sequential(*modules)
```
Comment (Pupy101) [inline src/model/models.py:25]: Переименовать в просто model и не забыть чекнуть в forward

File: src/model/models.py [hunk 2]
```diff
@@ -0,0 +1,235 @@
...
+        Returns:
+            out (torch.Tensor): Encoded image
+        """
+        batch_size = images.size(0)
+        out = self.efficientnet(images)  # (batch_size, 1280, 7, 7)
+        overal = self.avg_pool(out).permute(0, 2, 3, 1) # (batch_size, 1, 1, 1280)
+        out = self.adaptive_pool(out)  # (batch_size, 1280, count_windows, count_windows)
+        out = out.permute(0, 2, 3, 1).reshape(batch_size, -1, 1280)  # (batch_size, count_windows * count_windows, 1280)
+        
+        return torch.cat([out, overal], dim=2)
```
Comment (Pupy101) [inline src/model/models.py:49]: dim=1

File: src/model/models.py [hunk 3]
```diff
@@ -0,0 +1,235 @@
...
+        Initialization
+
+        Args:
+            encoder_dim (torch.Tensor): feature size of encoded images
+            decoder_dim (torch.Tensor): size of decoder's RNN
+        """
+        super(Attention, self).__init__()
+        self.wq = nn.Linear(decoder_dim, decoder_dim, bias=False)  # create Tensor Query
+        self.wk = nn.Linear(encoder_dim, decoder_dim, bias=False)  # create Tensor Key
+        self.wv = nn.Linear(decoder_dim, decoder_dim, bias=False)  # create Tensor Value
```
Comment (Pupy101) [inline src/model/models.py:80]: Wq, Wk, Wq ?

File: src/model/models.py [hunk 4]
```diff
@@ -0,0 +1,235 @@
...
+        attention_out = energy @ value  # (batch_size, seq_len, decoder_dim)
+        
+        return attention_out
+
+class ImageDecoder(nn.Module):
+    """
+    NLP model. Сreates an image description based on the attention model.
+    """
+    
+    def __init__(self, attention_dim, embed_dim, decoder_dim, vocab_size, encoder_dim=1280, dropout=0.5):
```
Comment (Pupy101) [inline src/model/models.py:109]: добавить аннотации типов. все должны быть int, поправить и docstring

File: src/model/models.py [hunk 5]
```diff
@@ -0,0 +1,235 @@
...
+        for c in list(self.resnet.children())[5:]:
+            for p in c.parameters():
+                p.requires_grad = fine_tune
+
+class Attention(nn.Module):
+    """
+    Attention model. Creates the parts that the nlp model should pay attention to.
+    """
+
+    def __init__(self, encoder_dim: torch.Tensor, decoder_dim: torch.Tensor, p: tp.Optional[float] = 0.2):
```
Comment (Pupy101) [inline src/model/models.py:69]: должны быть int, а не тензоры

File: src/model/models.py [hunk 6]
```diff
@@ -0,0 +1,235 @@
...
+        super().__init__()
+
+        self.encoder_dim = encoder_dim
+        self.attention_dim = attention_dim
+        self.embed_dim = embed_dim
+        self.decoder_dim = decoder_dim
+        self.vocab_size = vocab_size
+        self.dropout = dropout
+
+        self.attention = Attention(encoder_dim, decoder_dim, attention_dim)  # attention network
```
Comment (Pupy101) [inline src/model/models.py:130]: посмотреть параметры аттеншена и изменить, убрать из инита attention_dim

File: src/model/models.py [hunk 7]
```diff
@@ -0,0 +1,235 @@
...
+
+        self.embedding = nn.Embedding(vocab_size, embed_dim)  # embedding layer
+        self.dropout = nn.Dropout(p=self.dropout)
```
Comment (Pupy101) [inline src/model/models.py:133]: исправить self.dropout = nn.Dropout(p=p)

File: src/model/models.py [hunk 8]
```diff
@@ -0,0 +1,235 @@
...
+            vocab_size ([type]): size of vocabulary
+            encoder_dim (int, optional): feature size of encoded images. Defaults to 1280.
+            dropout (float, optional): dropout. Defaults to 0.5.
+        """
+        super().__init__()
+
+        self.encoder_dim = encoder_dim
+        self.attention_dim = attention_dim
+        self.embed_dim = embed_dim
+        self.decoder_dim = decoder_dim
```
Comment (Pupy101) [inline src/model/models.py:126]: вроде как не нужно

File: src/model/models.py [hunk 9]
```diff
@@ -0,0 +1,235 @@
...
+        self.vocab_size = vocab_size
+        self.dropout = dropout
```
Comment (Pupy101) [inline src/model/models.py:128]: тоже не нужно

File: src/model/models.py [hunk 10]
```diff
@@ -0,0 +1,235 @@
...
+        c = self.init_c(mean_encoder_out)
+        return h, c
+    
+    """
+        Forward propagation.
+        :param encoder_out: encoded images, a tensor of dimension (batch_size, enc_image_size, enc_image_size, encoder_dim)
+        :param encoded_captions: encoded captions, a tensor of dimension (batch_size, max_caption_length)
+        :param caption_lengths: caption lengths, a tensor of dimension (batch_size, 1)
+        :return: scores for vocabulary, sorted encoded captions, decode lengths, weights, sort indices
+        """
```
Comment (Pupy101) [inline src/model/models.py:174]: Убрать, в ините оставить слои, а среднее по изображению в forward - encoder_out[:, -1, :]

File: src/model/models.py [hunk 11]
```diff
@@ -0,0 +1,235 @@
...
+        self.attention = Attention(encoder_dim, decoder_dim, attention_dim)  # attention network
+
+        self.embedding = nn.Embedding(vocab_size, embed_dim)  # embedding layer
+        self.dropout = nn.Dropout(p=self.dropout)
+        self.decode_step = nn.LSTMCell(embed_dim + encoder_dim, decoder_dim, bias=True)  # decoding LSTMCell
+        self.init_h = nn.Linear(encoder_dim, decoder_dim)  # linear layer to find initial hidden state of LSTMCell
+        self.init_c = nn.Linear(encoder_dim, decoder_dim)  # linear layer to find initial cell state of LSTMCell
+        self.f_beta = nn.Linear(decoder_dim, encoder_dim)  # linear layer to create a sigmoid-activated gate
+        self.sigmoid = nn.Sigmoid()
+        self.fc = nn.Linear(decoder_dim, vocab_size)  # linear layer to find scores over vocabulary
```
Comment (Pupy101) [inline src/model/models.py:139]: лучше назвать lm - language model

File: src/model/models.py [hunk 12]
```diff
@@ -0,0 +1,235 @@
...
+        self.init_weights()  # initialize some layers with the uniform distribution
```
Comment (Pupy101) [inline src/model/models.py:140]: Ты кто такой сука!!! init_weight()

File: src/model/models.py [hunk 13]
```diff
@@ -0,0 +1,235 @@
...
+
+        self.attention = Attention(encoder_dim, decoder_dim, attention_dim)  # attention network
+
+        self.embedding = nn.Embedding(vocab_size, embed_dim)  # embedding layer
+        self.dropout = nn.Dropout(p=self.dropout)
+        self.decode_step = nn.LSTMCell(embed_dim + encoder_dim, decoder_dim, bias=True)  # decoding LSTMCell
+        self.init_h = nn.Linear(encoder_dim, decoder_dim)  # linear layer to find initial hidden state of LSTMCell
+        self.init_c = nn.Linear(encoder_dim, decoder_dim)  # linear layer to find initial cell state of LSTMCell
+        self.f_beta = nn.Linear(decoder_dim, encoder_dim)  # linear layer to create a sigmoid-activated gate
+        self.sigmoid = nn.Sigmoid()
```
Comment (Pupy101) [inline src/model/models.py:138]: не нужно

File: src/model/models.py [hunk 14]
```diff
@@ -0,0 +1,235 @@
...
+        self.fc = nn.Linear(decoder_dim, vocab_size)  # linear layer to find scores over vocabulary
+        self.init_weights()  # initialize some layers with the uniform distribution
+    
+    def init_weights(self):
+        """
+        Initializes some parameters with values from the uniform distribution, for easier convergence.
+        """
+        self.embedding.weight.data.uniform_(-0.1, 0.1)
+        self.fc.bias.data.fill_(0)
+        self.fc.weight.data.uniform_(-0.1, 0.1)
```
Comment (Pupy101) [inline src/model/models.py:148]: del

File: src/model/models.py [hunk 15]
```diff
@@ -0,0 +1,235 @@
...
+    def init_hidden_state(self, encoder_out):
+        """
+        Creates the initial hidden and cell states for the decoder's LSTM based on the encoded images.
+        :param encoder_out: encoded images, a tensor of dimension (batch_size, num_pixels, encoder_dim)
+        :return: hidden state, cell state
+        """
+        mean_encoder_out = encoder_out.mean(dim=1)
+        h = self.init_h(mean_encoder_out)  # (batch_size, decoder_dim)
+        c = self.init_c(mean_encoder_out)
+        return h, c
```
Comment (Pupy101) [inline src/model/models.py:166]: del

File: src/model/models.py [hunk 16]
```diff
@@ -0,0 +1,235 @@
...
+            alphas
+            sort_ind: 
+        """
+
+        batch_size = encoder_out.size(0)
+        encoder_dim = encoder_out.size(-1)
+        vocab_size = self.vocab_size
+
+        # Flatten image
+        encoder_out = encoder_out.view(batch_size, -1, encoder_dim)  # (batch_size, num_pixels, encoder_dim)
```
Comment (Pupy101) [inline src/model/models.py:197]: не надо, уже в image model сделано

File: src/model/models.py [hunk 17]
```diff
@@ -0,0 +1,235 @@
...
+        return h, c
+    
+    """
+        Forward propagation.
+        :param encoder_out: encoded images, a tensor of dimension (batch_size, enc_image_size, enc_image_size, encoder_dim)
+        :param encoded_captions: encoded captions, a tensor of dimension (batch_size, max_caption_length)
+        :param caption_lengths: caption lengths, a tensor of dimension (batch_size, 1)
+        :return: scores for vocabulary, sorted encoded captions, decode lengths, weights, sort indices
+        """
+    def forward(self, encoder_out: torch.Tensor, encoded_captions, caption_lengths):
```
Comment (Pupy101) [inline src/model/models.py:175]: добавить два аргумента: - p_teaher_forcing и убрать caption_lengths, типы добавить

File: src/model/models.py [hunk 18]
```diff
@@ -0,0 +1,235 @@
...
+        # Embedding
+        embeddings = self.embedding(encoded_captions)  # (batch_size, max_caption_length, embed_dim)
+
+        # Initialize LSTM state
+        h, c = self.init_hidden_state(encoder_out)  # (batch_size, decoder_dim)
+
+        # We won't decode at the <end> position, since we've finished generating as soon as we generate <end>
+        # So, decoding lengths are actual lengths - 1
+        decode_lengths = (caption_lengths - 1).tolist()
+
```
Comment (Pupy101) [inline src/model/models.py:214]: добавить teacher forcing - с вероятностью подавать правильный эмбеддинг - переменная embeddings и убрать ненужные слои.
1. Инициализируем h и с - линейными слоями и вектором encoder_out[:, -1, :]
2. убрать сортировку
3. Просто генерируем токен из - h, c, input- конкатенация ембеддинга и тензора из аттеншена.
Разобраться какие эмбеддинги нужно подавать в сеть с TF и без него.

File: src/model/models.py [hunk 19]
```diff
@@ -0,0 +1,235 @@
...
+            gate = self.sigmoid(self.f_beta(h[:batch_size_t]))  # gating scalar, (batch_size_t, encoder_dim)
+            attention_weighted_encoding = gate * attention_weighted_encoding
+            h, c = self.decode_step(
+                torch.cat([embeddings[:batch_size_t, t, :], attention_weighted_encoding], dim=1),
+                (h[:batch_size_t], c[:batch_size_t]))  # (batch_size_t, decoder_dim)
+            preds = self.fc(self.dropout(h))  # (batch_size_t, vocab_size)
+            predictions[:batch_size_t, t, :] = preds
+            alphas[:batch_size_t, t, :] = alpha
+
+        return predictions, encoded_captions, decode_lengths, alphas, sort_ind
```
Comment (Pupy101) [inline src/model/models.py:235]: убрать ненужно, оставить только predictions

**Review by Pupy101 — APPROVED** [review]

File: src/model/models.py [hunk 20]
```diff
@@ -0,0 +1,251 @@
...
+class ImageEncoder(nn.Module):
+    """
+    Vision model. Encodes the image for attention model.
+    """
+    def __init__(
+        self, 
+        model: tp.Optional[nn.Module] = models.efficientnet_b0(pretrained=True),
+        count_windows: tp.Optional[int] = 4, 
+        last_layer_for_del: tp.Optional[int] = 2,
+        feature_map: tp.Optional[int] = 7,
```
Comment (Pupy101) [inline src/model/models.py:18]: _size?

File: src/model/models.py [hunk 21]
```diff
@@ -0,0 +1,251 @@
...
+        query = self.dropout(self.wq(decoder_hidden))  # (batch_size, seq_len, decoder_dim)
+
+        key = self.dropout(self.wk(encoder_out))  # (batch_size, num_windows, decoder_dim)
+        value = self.dropout(self.wv(encoder_out))  # (batch_size, num_windows, decoder_dim)
+        energy = self.softmax(query @ key.t() / self.d)  # (batch_size, seq_len, num_windows)
+        attention_out = energy @ value  # (batch_size, seq_len, decoder_dim)
+        
+        return attention_out
+
+class ImageDecoder(nn.Module):
```
Comment (Pupy101) [inline src/model/models.py:120]: Как для проекта это можно использовать, но для лучшего результата можно использовать https://huggingface.co/docs/transformers/model_doc/gpt2#transformers.GPT2LMHeadModel
